tests/shadowcasting_test.cpp

```diff
@@ -412,197 +413,406 @@ static void shadowcasting_3d_2d( const int iterations )
+        caches[z] = new level_cache();
+        seen_squares[z] = &caches[z]->seen_cache;
+        transparency_cache[z] = &caches[z]->transparency_cache;
+        floor_cache[z] = &caches[z]->floor_cache;
+        for( int y = 0; y < MAPSIZE * SEEY; ++y ) {
+            for( int x = 0; x < MAPSIZE * SEEX; ++x ) {
+                caches[z]->transparency_cache[x][y] = test_case.get_global( x, y, z );
+                caches[z]->floor_cache[x][y] = test_case.get_floor( x, y, z );
+            }
         }
     }
 
-    castLightAll<float, float, sight_calc, sight_check, update_light, accumulate_transparency>(
-        seen_squares, transparency_cache, ORIGIN.x, ORIGIN.y );
+    if( fov_3d ) {
+        cast_zlight<float, sight_calc, sight_check, accumulate_transparency>( seen_squares,
+                transparency_cache, floor_cache, { ORIGIN.x, ORIGIN.y, ORIGIN.z - OVERMAP_DEPTH }, 0, 1.0 );
+    } else {
+        castLightAll<float, float, sight_calc, sight_check, update_light, accumulate_transparency>(
+            *seen_squares[11], *transparency_cache[11], ORIGIN.x, ORIGIN.y );
 
-    // Compares the whole grid, but out-of-bounds compares will de-facto pass.
-    for( int y = 0; y < expected_result.height(); ++y ) {
-        for( int x = 0; x < expected_result.width(); ++x ) {
-            INFO( "x:" << x << " y:" << y << " expected:" << expected_result.data[y][x] << " actual:" <<
-                  seen_squares[expected_result.offset.y + y][expected_result.offset.x + x] );
-            if( V == expected_result.get_local( x, y ) ) {
-                CHECK( seen_squares[expected_result.offset.y + y][expected_result.offset.x + x] > 0 );
-            } else {
-                CHECK( seen_squares[expected_result.offset.y + y][expected_result.offset.x + x] == 0 );
+    }
+    bool passed = true;
+    std::ostringstream trans_grid;
+    std::ostringstream expected_grid;
+    std::ostringstream actual_grid;
+    for( int gz = expected.offset.z; gz < expected.get_max().z; ++gz ) {
+        for( int gy = expected.offset.y; gy < expected.get_max().y; ++gy ) {
+            for( int gx = expected.offset.x; gx < expected.get_max().x; ++gx ) {
+                trans_grid << caches[gz]->transparency_cache[gx][gy];
+                expected_grid << ( expected.get_global( gx, gy, gz ) > 0 ? 'V' : 'O' );
+                actual_grid << ( ( *seen_squares[gz] )[gx][gy] > 0 ? 'V' : 'O' );
+                if( V == expected.get_global( gx, gy, gz ) && ( *seen_squares[gz] )[gx][gy] == 0 ) {
+                    passed = false;
+                } else if( O == expected.get_global( gx, gy, gz ) &&
+                           ( *seen_squares[gz] )[gx][gy] > 0 ) {
+                    passed = false;
+                }
             }
+            trans_grid << '\n';
+            expected_grid << '\n';
+            actual_grid << '\n';
         }
+        trans_grid << '\n';
+        expected_grid << '\n';
+        actual_grid << '\n';
+        delete caches[gz];
```

Comment:
Here's your leak, it allocates either one or 21 cache levels depending on FoV flag, but only deletes the ones that have a corresponding test case entry.

Reply:
Perfect thanks, for some reason my leak profiler was showing some memory leakage on loading a normal game so I wasn’t looking at the tests.

Reply:
And yes I have realized now that asan totally gave the function name for where the leak was, I was running on very little sleep when I first looked at this.